File: scripts/build_favicon.py
"""Render the site icon: the real Earth, with one white flash on it.

The globe on the page samples ``public/earth/earth_day_4096.jpg``; so does this, so the icon is
the same planet the reader is about to watch rather than a drawing of one. Each output pixel is
inverse-projected off an orthographic sphere, sampled bilinearly out of that equirectangular
texture, shaded, and composited under a white flash.

Writes, all of which Next's App Router picks up by filename:

    app/favicon.ico      16, 32, 48 — the bare /favicon.ico browsers ask for unprompted
    app/icon.png         512 — the one referenced by <link rel="icon">
    app/apple-icon.png   180 — home-screen bookmarks

Run with ``pnpm run build:favicon``. Pillow is a dev dependency group in pyproject.toml.
"""

from __future__ import annotations

import math
from pathlib import Path

from PIL import Image, ImageEnhance

ROOT = Path(__file__).resolve().parent.parent
TEXTURE = ROOT / "public" / "earth" / "earth_day_4096.jpg"
APP = ROOT / "app"

# Where the globe faces. Africa and Europe centred: the most recognisable hemisphere at a size
# where three or four coastlines is all the detail that survives, and the one the story's own
# maps open on.
VIEW_LON = 14.0
VIEW_LAT = 16.0

# Sunlight direction in view space, as a unit-ish vector. Up and slightly left, so the terminator
# falls across the lower right and the disc reads as a sphere instead of a sticker.
LIGHT = (-0.34, -0.46, 0.82)
# How dark the night side gets. Not black: an unlit half that goes to zero looks like a bite out
# of the icon at 16px rather than a planet turning away.
NIGHT_FLOOR = 0.30
# Thin blue atmosphere on the limb, which is most of what separates the disc from a dark tab.
ATMOSPHERE = (150, 195, 255)

# The flash: one death, in white. Over the Congo basin rather than anywhere prettier — it is the
# darkest green on the visible hemisphere, and at 16px contrast against the ground underneath is
# the only thing keeping the flash from dissolving into the Sahara.
FLASH_LON = 20.0
FLASH_LAT = 1.0
FLASH_CORE = 0.048  # solid white, in units of the icon box
FLASH_GLOW = 0.165  # falloff around it

DISC_R = 0.465  # planet radius, in units of the icon box
SUPERSAMPLE = 4
# Below this, a faithful render is a muddy one: sixteen pixels of Earth is mostly the average of
# a coastline, and it needs the push to still read as land against water.
SMALL = 64
SMALL_SATURATION = 1.22
SMALL_CONTRAST = 1.12
# The flash is optically scaled rather than geometrically faithful. A dot that is right at 512px
# is a fifth of a pixel at 16 and disappears; the icon's job at that size is to say "Earth, and
# something happened on it", which needs the mark bigger than the projection would put it.
SMALL_FLASH_BOOST = 1.15

Rgb = tuple[float, float, float]


def _clamp01(v: float) -> float:
    return 0.0 if v < 0.0 else 1.0 if v > 1.0 else v


def _rotation(view_lon: float, view_lat: float):
    """View-space (x right, y down, z toward the viewer) to unit sphere, as three basis rows."""
    lon = math.radians(view_lon)
    lat = math.radians(view_lat)
    # Sphere point for the centre of the disc, and the two tangents that span the view plane.
    forward = (math.cos(lat) * math.sin(lon), math.sin(lat), math.cos(lat) * math.cos(lon))
    right = (math.cos(lon), 0.0, -math.sin(lon))
    up = (
        -math.sin(lat) * math.sin(lon),
        math.cos(lat),
        -math.sin(lat) * math.cos(lon),
    )
    return right, up, forward


def _sphere_to_uv(p: tuple[float, float, float]) -> tuple[float, float]:
    """Unit sphere point to equirectangular texture coordinates in [0, 1]."""
    x, y, z = p
    lon = math.atan2(x, z)
    lat = math.asin(max(-1.0, min(1.0, y)))
    return (lon / (2 * math.pi) + 0.5) % 1.0, 0.5 - lat / math.pi


class Texture:
    """Bilinear sampler over the equirectangular day map, wrapping in longitude."""

    def __init__(self, path: Path, width: int) -> None:
        source = Image.open(path).convert("RGB")
        # Downsampled first: the icon is never larger than 512, and averaging 4096 px of coastline
        # down to that in one Lanczos pass is both faster and cleaner than point-sampling it.
        self.image = source.resize((width, width // 2), Image.LANCZOS)
        self.w, self.h = self.image.size
        self.px = self.image.load()

    def sample(self, u: float, v: float) -> Rgb:
        x = u * self.w - 0.5
        y = _clamp01(v) * self.h - 0.5
        x0 = math.floor(x)
        y0 = max(0, min(self.h - 1, math.floor(y)))
        fx = x - x0
        fy = y - y0
        x0 %= self.w
        x1 = (x0 + 1) % self.w
        y1 = min(self.h - 1, y0 + 1)
        c00 = self.px[x0, y0]
        c10 = self.px[x1, y0]
        c01 = self.px[x0, y1]
        c11 = self.px[x1, y1]
        return tuple(  # type: ignore[return-value]
            (c00[i] * (1 - fx) + c10[i] * fx) * (1 - fy) + (c01[i] * (1 - fx) + c11[i] * fx) * fy
            for i in range(3)
        )


def _flash_offset(right, up, forward) -> tuple[float, float] | None:
    """Where the flash lands on the disc, or None when it is round the back."""
    lon = math.radians(FLASH_LON)
    lat = math.radians(FLASH_LAT)
    p = (math.cos(lat) * math.sin(lon), math.sin(lat), math.cos(lat) * math.cos(lon))
    depth = sum(a * b for a, b in zip(p, forward))
    if depth <= 0.05:
        return None
    return (
        sum(a * b for a, b in zip(p, right)) * DISC_R,
        -sum(a * b for a, b in zip(p, up)) * DISC_R,
    )


def render(size: int, texture: Texture) -> Image.Image:
    right, up, forward = _rotation(VIEW_LON, VIEW_LAT)
    flash = _flash_offset(right, up, forward)
    boost = 1.0 if size >= SMALL else 1.0 + (SMALL - size) / SMALL * SMALL_FLASH_BOOST
    core = FLASH_CORE * boost
    halo_r = FLASH_GLOW * boost
    ss = SUPERSAMPLE
    big = size * ss
    out = Image.new("RGBA", (big, big), (0, 0, 0, 0))
    px = out.load()

    for py in range(big):
        # View-space coordinates of this sample, in units of the icon box, origin at the centre.
        vy = (py + 0.5) / big - 0.5
        for pxi in range(big):
            vx = (pxi + 0.5) / big - 0.5
            dist = math.hypot(vx, vy)
            if dist > DISC_R:
                continue

            # Back-project onto the sphere: x and y are the view plane, z falls out of the radius.
            nx = vx / DISC_R
            ny = -vy / DISC_R
            nz = math.sqrt(max(0.0, 1.0 - nx * nx - ny * ny))
            point = tuple(nx * right[i] + ny * up[i] + nz * forward[i] for i in range(3))
            r, g, b = texture.sample(*_sphere_to_uv(point))  # type: ignore[arg-type]

            # Day/night. The floor keeps the unlit half readable; the gamma softens the terminator
            # so it does not band across four pixels at icon sizes.
            lambert = nx * LIGHT[0] + (-ny) * LIGHT[1] + nz * LIGHT[2]
            shade = NIGHT_FLOOR + (1.0 - NIGHT_FLOOR) * _clamp01(lambert * 1.25 + 0.22) ** 0.75
            r, g, b = r * shade, g * shade, b * shade

            # Atmospheric limb, strongest on the lit edge.
            edge = _clamp01((dist - DISC_R * 0.82) / (DISC_R * 0.18)) ** 2
            glow = edge * (0.35 + 0.5 * _clamp01(lambert + 0.35))
            r = r + (ATMOSPHERE[0] - r) * glow * 0.65
            g = g + (ATMOSPHERE[1] - g) * glow * 0.65
            b = b + (ATMOSPHERE[2] - b) * glow * 0.65

            if flash is not None:
                fd = math.hypot(vx - flash[0], vy - flash[1])
                if fd < halo_r:
                    halo = (1.0 - fd / halo_r) ** 2
                    r = r + (255 - r) * halo * 0.92
                    g = g + (255 - g) * halo * 0.92
                    b = b + (255 - b) * halo * 0.92
                if fd < core:
                    r = g = b = 255.0

            px[pxi, py] = (round(_clamp01(r / 255) * 255), round(_clamp01(g / 255) * 255),
                           round(_clamp01(b / 255) * 255), 255)

    out = out.resize((size, size), Image.LANCZOS)
    if size <= SMALL:
        out = ImageEnhance.Color(out).enhance(SMALL_SATURATION)
        out = ImageEnhance.Contrast(out).enhance(SMALL_CONTRAST)
    return out


def main() -> None:
    if not TEXTURE.exists():
        raise SystemExit(f"missing Earth texture: {TEXTURE}")

    texture = Texture(TEXTURE, 2048)
    sizes = [16, 32, 48, 180, 512]
    rendered = {size: render(size, texture) for size in sizes}

    # Pillow writes a multi-size ICO from one image plus a size list, resampling internally; each
    # size is rendered from the texture instead, so the 16px one is a real 16px render rather than
    # a 512px one squeezed down to mud.
    ico = APP / "favicon.ico"
    rendered[48].save(
        ico,
        format="ICO",
        sizes=[(16, 16), (32, 32), (48, 48)],
        append_images=[rendered[16], rendered[32]],
    )
    rendered[512].save(APP / "icon.png", format="PNG", optimize=True)
    rendered[180].save(APP / "apple-icon.png", format="PNG", optimize=True)
    print(f"Wrote {ico.name} (16, 32, 48), icon.png (512) and apple-icon.png (180)")


if __name__ == "__main__":
    main()

File: scripts/build-favicon.ts
// Draws the site's icon — a night-side planet with one flash on it — and writes both forms of
// it: app/icon.svg for anything modern, app/favicon.ico for the bare `/favicon.ico` request
// browsers still make on their own.
//
// The geometry is here rather than in a checked-in binary so the two files cannot drift: the SVG
// and the raster are the same numbers, and regenerating is `pnpm run build:favicon`.
//
// No image dependency. The ICO container is a header plus one bottom-up BGRA bitmap per size,
// which is about forty lines, and pulling in an encoder to avoid them would be the larger cost.
import { writeFileSync } from "node:fs";
import { join } from "node:path";

// Everything below is in units of the icon's own box, so one set of numbers serves every size.
const R = 0.46; // planet radius
const CX = 0.5;
const CY = 0.5;
// Where the light comes from. The lit limb runs along this direction; the rest is night side.
const LIGHT = { x: -0.62, y: -0.78 };
// The flash: one death, on the dark side where it reads. Its glow is what survives 16px.
const FLASH = { x: 0.6, y: 0.38, r: 0.078, glow: 0.2 };

const NIGHT = [11, 14, 38] as const; // #0b0e26 — the story's pre-hero sky, a shade deeper
const DAY = [58, 86, 168] as const; // #3a56a8 — the lit crescent
const LIMB = [150, 183, 255] as const; // #96b7ff — the rim light that separates disc from page
const FLASH_RGB = [255, 122, 74] as const; // #ff7a4a — the same warm accent the globe flashes

type Rgba = [number, number, number, number];

const clamp01 = (v: number) => (v < 0 ? 0 : v > 1 ? 1 : v);
const mix = (a: readonly number[], b: readonly number[], t: number) =>
  a.map((v, i) => v + ((b[i] as number) - v) * t);

// Colour of one point inside the unit box, or transparent outside the disc. Antialiasing is done
// by the caller through supersampling, so this is a hard-edged function.
function shade(x: number, y: number): Rgba {
  const dx = x - CX;
  const dy = y - CY;
  const dist = Math.hypot(dx, dy);
  if (dist > R) return [0, 0, 0, 0];

  // How much this point faces the light, as the dot product of the surface normal with the light
  // direction. Treating the disc as a hemisphere is what stops the icon reading as a flat coin.
  const nz = Math.sqrt(Math.max(0, R * R - dist * dist)) / R;
  const nx = dx / R;
  const ny = dy / R;
  const lz = Math.sqrt(Math.max(0, 1 - LIGHT.x * LIGHT.x - LIGHT.y * LIGHT.y));
  const lambert = clamp01(nx * LIGHT.x + ny * LIGHT.y + nz * lz);
  // Biased towards night: this is a planet at dusk, not a lamp.
  let rgb = mix(NIGHT, DAY, clamp01(lambert * 1.35 - 0.15));

  // Rim light on the lit limb. Wider and brighter than a physical terminator would be: on a
  // dark browser tab the disc is nearly the colour of the chrome, and this arc is the only thing
  // that says there is a planet there at all.
  const edge = clamp01((dist - R * 0.78) / (R * 0.22));
  rgb = mix(rgb, LIMB, edge * edge * clamp01(lambert * 1.9 + 0.06));

  // The flash, and the halo that makes it visible once the whole icon is sixteen pixels wide.
  const fd = Math.hypot(x - CX - FLASH.x * R, y - CY - FLASH.y * R);
  if (fd < FLASH.glow) {
    const halo = 1 - fd / FLASH.glow;
    rgb = mix(rgb, FLASH_RGB, halo * halo * 0.75);
  }
  if (fd < FLASH.r) rgb = [...FLASH_RGB];

  return [rgb[0] as number, rgb[1] as number, rgb[2] as number, 255];
}

// One size, supersampled. 4×4 per pixel is enough for a disc and costs nothing at these sizes.
function raster(size: number): Uint8Array {
  const SS = 4;
  const out = new Uint8Array(size * size * 4);
  for (let py = 0; py < size; py++) {
    for (let px = 0; px < size; px++) {
      let r = 0;
      let g = 0;
      let b = 0;
      let a = 0;
      for (let sy = 0; sy < SS; sy++) {
        for (let sx = 0; sx < SS; sx++) {
          const [pr, pg, pb, pa] = shade(
            (px + (sx + 0.5) / SS) / size,
            (py + (sy + 0.5) / SS) / size,
          );
          // Premultiplied, so a transparent sample cannot bleed black into the edge.
          const w = pa / 255;
          r += pr * w;
          g += pg * w;
          b += pb * w;
          a += pa;
        }
      }
      const n = SS * SS;
      const alpha = a / n;
      const scale = alpha > 0 ? 255 / a : 0;
      const i = (py * size + px) * 4;
      out[i] = Math.round(r * scale);
      out[i + 1] = Math.round(g * scale);
      out[i + 2] = Math.round(b * scale);
      out[i + 3] = Math.round(alpha);
    }
  }
  return out;
}

// One ICO image: a BITMAPINFOHEADER, the pixels bottom-up as BGRA, then an empty AND mask (the
// alpha channel already carries the transparency, but the mask is not optional in the format).
function icoImage(size: number, rgba: Uint8Array): Buffer {
  const header = Buffer.alloc(40);
  header.writeUInt32LE(40, 0); // biSize
  header.writeInt32LE(size, 4); // biWidth
  header.writeInt32LE(size * 2, 8); // biHeight — XOR bitmap plus AND mask
  header.writeUInt16LE(1, 12); // biPlanes
  header.writeUInt16LE(32, 14); // biBitCount
  const maskStride = Math.ceil(size / 32) * 4;
  const xor = Buffer.alloc(size * size * 4);
  const and = Buffer.alloc(maskStride * size);
  header.writeUInt32LE(xor.length + and.length, 20); // biSizeImage

  for (let y = 0; y < size; y++) {
    const src = (size - 1 - y) * size * 4; // rows run bottom-up
    for (let x = 0; x < size; x++) {
      const s = src + x * 4;
      const d = (y * size + x) * 4;
      xor[d] = rgba[s + 2] as number; // B
      xor[d + 1] = rgba[s + 1] as number; // G
      xor[d + 2] = rgba[s] as number; // R
      xor[d + 3] = rgba[s + 3] as number; // A
    }
  }
  return Buffer.concat([header, xor, and]);
}

function ico(sizes: number[]): Buffer {
  const images = sizes.map((size) => icoImage(size, raster(size)));
  const dir = Buffer.alloc(6 + images.length * 16);
  dir.writeUInt16LE(0, 0);
  dir.writeUInt16LE(1, 2); // 1 = icon
  dir.writeUInt16LE(images.length, 4);
  let offset = dir.length;
  images.forEach((image, i) => {
    const at = 6 + i * 16;
    const size = sizes[i] as number;
    dir.writeUInt8(size >= 256 ? 0 : size, at);
    dir.writeUInt8(size >= 256 ? 0 : size, at + 1);
    dir.writeUInt16LE(1, at + 4); // planes
    dir.writeUInt16LE(32, at + 6); // bits per pixel
    dir.writeUInt32LE(image.length, at + 8);
    dir.writeUInt32LE(offset, at + 12);
    offset += image.length;
  });
  return Buffer.concat([dir, ...images]);
}

// The same planet as vector, for browsers that prefer it and for anywhere the icon is shown big.
function svg(): string {
  const pct = (v: number) => (v * 100).toFixed(2);
  const rgb = (c: readonly number[]) => `rgb(${c[0]} ${c[1]} ${c[2]})`;
  return `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 64 64" role="img" aria-label="A planet with one flash on its night side">
  <defs>
    <radialGradient id="face" cx="${pct(CX + (LIGHT.x * R) / 2)}%" cy="${pct(CY + (LIGHT.y * R) / 2)}%" r="78%">
      <stop offset="0%" stop-color="${rgb(DAY)}" />
      <stop offset="62%" stop-color="${rgb(NIGHT)}" />
      <stop offset="100%" stop-color="${rgb(NIGHT)}" />
    </radialGradient>
    <radialGradient id="flash" cx="50%" cy="50%" r="50%">
      <stop offset="0%" stop-color="${rgb(FLASH_RGB)}" />
      <stop offset="35%" stop-color="${rgb(FLASH_RGB)}" stop-opacity="0.75" />
      <stop offset="100%" stop-color="${rgb(FLASH_RGB)}" stop-opacity="0" />
    </radialGradient>
  </defs>
  <circle cx="32" cy="32" r="${(R * 64).toFixed(2)}" fill="url(#face)" />
  <circle cx="32" cy="32" r="${(R * 64 - 0.7).toFixed(2)}" fill="none" stroke="${rgb(LIMB)}" stroke-opacity="0.85" stroke-width="1.4" stroke-dasharray="34 58" stroke-dashoffset="20" transform="rotate(-142 32 32)" />
  <circle cx="${(32 + FLASH.x * R * 64).toFixed(2)}" cy="${(32 + FLASH.y * R * 64).toFixed(2)}" r="${(FLASH.glow * 64).toFixed(2)}" fill="url(#flash)" />
  <circle cx="${(32 + FLASH.x * R * 64).toFixed(2)}" cy="${(32 + FLASH.y * R * 64).toFixed(2)}" r="${(FLASH.r * 64).toFixed(2)}" fill="${rgb(FLASH_RGB)}" />
</svg>
`;
}

const app = join(process.cwd(), "app");
writeFileSync(join(app, "favicon.ico"), ico([16, 32, 48]));
writeFileSync(join(app, "icon.svg"), svg());
console.info("Wrote app/favicon.ico (16, 32, 48) and app/icon.svg");
